File: users/src/main/java/com/auth/users/services/UserService.java
package com.auth.users.services;

import com.auth.users.dtos.*;
import com.auth.users.entities.UserEntity;
import com.auth.users.entities.UserRole;
import com.auth.users.messaging.UserEventPublisher;
import com.auth.users.repositories.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class UserService {

    private final UserRepository userRepository;
    private final KeycloakService keycloakService;
    private final EventLookupService eventLookupService;
    private final UserEventPublisher userEventPublisher;

    @Transactional
    public UserProfileResponse registerUser(RegisterRequest registerRequest) {
        if (registerRequest.getUsername() == null || registerRequest.getUsername().trim().isEmpty()) {
            throw new IllegalArgumentException("Username is required");
        }
        if (registerRequest.getPassword() == null || registerRequest.getPassword().trim().isEmpty()) {
            throw new IllegalArgumentException("Password is required");
        }

        String username = registerRequest.getUsername();
        String email = registerRequest.getEmail();
        if (email == null || email.trim().isEmpty()) {
            email = username + "@example.com";
        }

        UserRole role = registerRequest.getRole();
        if (role == null) {
            role = UserRole.PARTICIPANT;
        }

        if (userRepository.existsByUsername(username)) {
            throw new IllegalArgumentException("Username already exists");
        }
        if (userRepository.existsByEmail(email)) {
            throw new IllegalArgumentException("Email already exists");
        }

        String roleName = role.name();
        String keycloakUserId = keycloakService.createUser(
                username,
                email,
                registerRequest.getPassword(),
                registerRequest.getFirstName(),
                registerRequest.getLastName(),
                roleName
        );

        // 2. Persist locally
        UserEntity userEntity = UserEntity.builder()
                .id(keycloakUserId)
                .username(username)
                .email(email)
                .firstName(registerRequest.getFirstName())
                .lastName(registerRequest.getLastName())
                .role(role)
                .build();

        UserEntity savedUser = userRepository.save(userEntity);
        log.info("Registered user {} with role {} locally and in Keycloak", savedUser.getUsername(), savedUser.getRole());

        // Fire-and-forget: broadcast a USER_CREATED event so the notification-service
        // can push a live toast to every connected frontend.
        userEventPublisher.publishUserCreated(savedUser);

        return mapToProfileResponse(savedUser);
    }

    public LoginResponse loginUser(LoginRequest loginRequest) {
        LoginResponse response = keycloakService.authenticate(loginRequest);

        UserEntity user = userRepository.findByUsername(loginRequest.getUsername())
                .orElseThrow(() -> new RuntimeException("User profile not found in local database"));

        response.setRole(user.getRole().name());
        return response;
    }

    public UserProfileResponse getUserProfile(String userId) {
        UserEntity user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        return mapToProfileResponse(user);
    }

    @Transactional
    public UserProfileResponse updateUserProfile(String userId, UserUpdateRequest updateRequest) {
        UserEntity user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));

        if (updateRequest.getFirstName() != null) {
            user.setFirstName(updateRequest.getFirstName());
        }
        if (updateRequest.getLastName() != null) {
            user.setLastName(updateRequest.getLastName());
        }

        UserEntity savedUser = userRepository.save(user);
        log.info("Updated user profile for user ID: {}", userId);
        return mapToProfileResponse(savedUser);
    }

    public List<UserProfileResponse> getAllUsers() {
        return userRepository.findAll().stream()
                .map(this::mapToProfileResponse)
                .collect(Collectors.toList());
    }

    /**
     * Returns every user that is allowed to organize events
     * (i.e. everyone whose role is {@code ORGANISATEUR} or {@code ADMINISTRATEUR}).
     * <p>
     * The returned {@link UserSummaryResponse} is purposely minimal — no email,
     * no timestamps — so this endpoint is safe to expose to any authenticated
     * user and can be used to populate an organizer-picker.
     */
    public List<UserSummaryResponse> getOrganizers() {
        return userRepository.findAll().stream()
                .filter(u -> u.getRole() == UserRole.ORGANISATEUR
                          || u.getRole() == UserRole.ADMINISTRATEUR)
                .map(this::mapToSummary)
                .collect(Collectors.toList());
    }

    private UserSummaryResponse mapToSummary(UserEntity user) {
        return UserSummaryResponse.builder()
                .id(user.getId())
                .username(user.getUsername())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .role(user.getRole())
                .build();
    }

    /**
     * Returns the events organized by the given user, joined from the events
     * microservice via Feign. If the user has no events (events-service returns
     * 404) or the events-service is unreachable, this returns an empty list.
     *
     * @throws RuntimeException if the local user does not exist.
     */
    public List<EventDto> getEventsForUser(String userId) {
        // Verify the user exists locally so we fail fast with a meaningful error
        // instead of returning an empty list for a non-existent user.
        if (!userRepository.existsById(userId)) {
            throw new RuntimeException("User not found: " + userId);
        }
        return eventLookupService.findEventsForUserSafe(userId);
    }

    /**
     * Lists every user together with the events they organize.
     * <p>
     * Uses a single call to {@code GET /api/events/all} on the events-service
     * (via Feign) and groups events per {@code organizerId} in memory, so this
     * costs O(N + M) with N = users and M = total events — no N+1 fan-out.
     */
    public List<UserWithEventsResponse> getAllUsersWithEvents() {
        List<UserEntity> users = userRepository.findAll();
        List<EventDto> allEvents = eventLookupService.findAllEventsSafe();

        Map<String, List<EventDto>> eventsByOrganizer = allEvents.stream()
                .filter(e -> e.getOrganizerId() != null)
                .collect(Collectors.groupingBy(EventDto::getOrganizerId));

        return users.stream()
                .map(u -> {
                    List<EventDto> events = eventsByOrganizer
                            .getOrDefault(u.getId(), java.util.Collections.emptyList());
                    return UserWithEventsResponse.builder()
                            .id(u.getId())
                            .username(u.getUsername())
                            .email(u.getEmail())
                            .firstName(u.getFirstName())
                            .lastName(u.getLastName())
                            .role(u.getRole())
                            .createdAt(u.getCreatedAt())
                            .eventCount(events.size())
                            .events(events)
                            .build();
                })
                .collect(Collectors.toList());
    }

    /**
     * Ensures a local UserEntity exists for the Keycloak user represented by the given JWT.
     * If the user does not exist locally yet (e.g. registered directly via Keycloak's
     * hosted registration page), this creates the local mirror record from JWT claims.
     * The local {@code id} always equals the Keycloak {@code sub}.
     */
    @Transactional
    @SuppressWarnings("unchecked")
    public UserProfileResponse syncFromJwt(Jwt jwt) {
        String userId = jwt.getSubject();

        return userRepository.findById(userId)
                .map(this::mapToProfileResponse)
                .orElseGet(() -> {
                    String username = jwt.getClaimAsString("preferred_username");
                    String email = jwt.getClaimAsString("email");
                    String firstName = jwt.getClaimAsString("given_name");
                    String lastName = jwt.getClaimAsString("family_name");

                    if (username == null || username.isBlank()) {
                        username = (email != null && !email.isBlank()) ? email : userId;
                    }
                    if (email == null || email.isBlank()) {
                        email = username + "@example.com";
                    }

                    UserRole role = extractRoleFromJwt(jwt);

                    UserEntity entity = UserEntity.builder()
                            .id(userId)
                            .username(username)
                            .email(email)
                            .firstName(firstName)
                            .lastName(lastName)
                            .role(role)
                            .build();

                    UserEntity saved = userRepository.save(entity);
                    log.info("Synced local profile for Keycloak user {} (role={})", saved.getUsername(), saved.getRole());

                    // First time we ever see this user locally — treat it as a "user created" event
                    // so users who register directly via Keycloak's hosted page still trigger a notification.
                    userEventPublisher.publishUserCreated(saved);

                    return mapToProfileResponse(saved);
                });
    }

    @SuppressWarnings("unchecked")
    private UserRole extractRoleFromJwt(Jwt jwt) {
        Map<String, Object> realmAccess = jwt.getClaim("realm_access");
        if (realmAccess != null && realmAccess.get("roles") instanceof Collection<?> roles) {
            for (Object role : roles) {
                String r = role.toString();
                for (UserRole ur : UserRole.values()) {
                    if (ur.name().equalsIgnoreCase(r)) {
                        return ur;
                    }
                }
            }
        }
        return UserRole.PARTICIPANT;
    }

    private UserProfileResponse mapToProfileResponse(UserEntity user) {
        return UserProfileResponse.builder()
                .id(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .role(user.getRole())
                .createdAt(user.getCreatedAt())
                .build();
    }
}

File: frontend/src/services/userService.ts
import api from '@/lib/api'

// Event type mirrors the Ms-Event Event entity as returned via users-service.
// Kept local to avoid a circular import with @/types/event (same shape).
export interface UserEvent {
  id: number
  title: string
  description: string
  category: string
  maxPlaces: number
  availablePlaces: number
  organizerId: string
  status: string
  schedules?: unknown[]
}

export type UserRole = 'ADMINISTRATEUR' | 'ORGANISATEUR' | 'PARTICIPANT'

export interface UserProfile {
  id: string
  username: string
  email: string
  firstName?: string
  lastName?: string
  role: UserRole
  createdAt?: string
}

/**
 * Lightweight user representation returned by GET /api/users/organizers.
 * Safe to display to any authenticated user (no email / timestamps).
 */
export interface UserSummary {
  id: string
  username: string
  firstName?: string
  lastName?: string
  role: UserRole
}

export interface UserUpdateRequest {
  firstName?: string
  lastName?: string
}

/**
 * Payload accepted by the admin-only user creation endpoint
 * (POST /api/users). Mirrors the backend RegisterRequest DTO.
 */
export interface AdminCreateUserRequest {
  username: string
  email?: string
  password: string
  firstName?: string
  lastName?: string
  role: UserRole
}

const USERS = '/api/users'

export const userService = {
  /** Current user profile from the local users-service DB. */
  getMyProfile: async (): Promise<UserProfile> => {
    const { data } = await api.get<UserProfile>(`${USERS}/profile`)
    return data
  },

  updateMyProfile: async (body: UserUpdateRequest): Promise<UserProfile> => {
    const { data } = await api.put<UserProfile>(`${USERS}/profile`, body)
    return data
  },

  /** Admin-only: list all users. */
  getAllUsers: async (): Promise<UserProfile[]> => {
    const { data } = await api.get<UserProfile[]>(USERS)
    return data
  },

  /**
   * Admin-only: create a new user with a specific role.
   * Reuses the same server flow as public registration but is guarded by
   * the ADMINISTRATEUR realm role. The backend also creates the account in
   * Keycloak and publishes a USER_CREATED event to RabbitMQ.
   */
  createUser: async (body: AdminCreateUserRequest): Promise<UserProfile> => {
    const { data } = await api.post<UserProfile>(USERS, body)
    return data
  },

  /**
   * Admin-only: delete a user from Keycloak + local DB.
   * The backend publishes a `USER_DELETED` event to RabbitMQ, which the
   * notification-service consumes and broadcasts to all connected frontends
   * (bell dropdown + `/notifications` page).
   */
  deleteUser: async (userId: string): Promise<void> => {
    await api.delete(`${USERS}/${userId}`)
  },

  /**
   * Admin-only: change a user's role.
   * Backend updates both Keycloak's realm-role mapping and the local DB,
   * then publishes a `USER_UPDATED` event to RabbitMQ.
   */
  changeUserRole: async (userId: string, role: UserRole): Promise<UserProfile> => {
    const { data } = await api.patch<UserProfile>(`${USERS}/${userId}/role`, { role })
    return data
  },

  /**
   * Lightweight list of users allowed to organize events (ORGANISATEUR + ADMINISTRATEUR).
   * Accessible to any authenticated user; used to populate the organizer picker
   * when creating an event.
   */
  getOrganizers: async (): Promise<UserSummary[]> => {
    const { data } = await api.get<UserSummary[]>(`${USERS}/organizers`)
    return data
  },

  /**
   * Events organized by the currently authenticated user.
   * Delegates to users-service GET /api/users/events, which itself calls the
   * events microservice via Feign (with JWT propagated).
   * <p>
   * Returns an empty array when the user organizes no events (users-service
   * normalizes the events-service 404 into an empty list on the backend).
   */
  getMyEvents: async (): Promise<UserEvent[]> => {
    const { data } = await api.get<UserEvent[]>(`${USERS}/events`)
    return data
  },

  /**
   * Ensure a local UserEntity exists for the currently authenticated Keycloak user.
   * Safe to call on every login. Returns the (possibly newly-created) local profile.
   */
  syncCurrentUser: async (): Promise<UserProfile> => {
    const { data } = await api.post<UserProfile>(`${USERS}/sync`)
    return data
  },
}
